eqcorrscan/lib/multi_corr.c

```diff
@@ -316,13 +373,180 @@ int normxcorr_time(float *template, int template_len, float *image, int image_le
+            free_fftw_arrays(i + 1, template_ext, image_ext, ccc, outa, outb, out);
+            return 1;
+        }
+
+        /* allocate image_ext arrays */
+        image_ext[i] = fftw_alloc_real(fft_len);
+        if (image_ext[i] == NULL) {
+            printf("Error allocating image_ext[%d]\n", i);
+            free_fftw_arrays(i + 1, template_ext, image_ext, ccc, outa, outb, out);
+            return 1;
+        }
+
+        /* allocate ccc arrays */
+        ccc[i] = fftw_alloc_real((long) fft_len * n_templates);
+        if (ccc[i] == NULL) {
+            printf("Error allocating ccc[%d]\n", i);
+            free_fftw_arrays(i + 1, template_ext, image_ext, ccc, outa, outb, out);
+            return 1;
+        }
+
+        /* allocate outa arrays */
+        outa[i] = fftw_alloc_complex((long) N2 * n_templates);
+        if (outa[i] == NULL) {
+            printf("Error allocating outa[%d]\n", i);
+            free_fftw_arrays(i + 1, template_ext, image_ext, ccc, outa, outb, out);
+            return 1;
+        }
+
+        /* allocate outb arrays */
+        outb[i] = fftw_alloc_complex((long) N2);
+        if (outb[i] == NULL) {
+            printf("Error allocating outb[%d]\n", i);
+            free_fftw_arrays(i + 1, template_ext, image_ext, ccc, outa, outb, out);
+            return 1;
+        }
+
+        /* allocate out arrays */
+        out[i] = fftw_alloc_complex((long) N2 * n_templates);
+        if (out[i] == NULL) {
+            printf("Error allocating out[%d]\n", i);
+            free_fftw_arrays(i + 1, template_ext, image_ext, ccc, outa, outb, out);
+            return 1;
+        }
+    }
+
+    //TODO: touch all arrays - NUMA first touch???
+
+    // We create the plans here since they are not thread safe.
+    pa = fftw_plan_dft_r2c_2d(n_templates, fft_len, template_ext[0], outa[0], FFTW_ESTIMATE);
+    pb = fftw_plan_dft_r2c_1d(fft_len, image_ext[0], outb[0], FFTW_ESTIMATE);
+    px = fftw_plan_dft_c2r_2d(n_templates, fft_len, out[0], ccc[0], FFTW_ESTIMATE);
+
+    /* loop over the channels */
+    #pragma omp parallel for reduction(+:r) num_threads(num_threads)
     for (i = 0; i < n_channels; ++i){
-        r = normxcorr_fftw_threaded(&templates[n_templates * template_len * i], template_len,
-                                    n_templates, &image[image_len * i], image_len,
-                                    &ccc[(image_len - template_len + 1) * n_templates * i], fft_len);
+        /* each thread has its own workspace */
+        int tid = 0;
+        #ifdef N_THREADS
+        /* get the id of this thread */
+        tid = omp_get_thread_num();
+        #endif
+
+        /* initialise memory to zero */
+        memset(template_ext[tid], 0, (long) fft_len * n_templates * sizeof(double));
+        memset(image_ext[tid], 0, (long) fft_len * sizeof(double));
+
+        /* call the routine */
+        r = normxcorr_fftw_main(&templates[(long) n_templates * template_len * i], template_len,
```

Comment:
We could probably leave it as it is for now, as it only seems to be an issue when computing the offsets, and think about changing later so that it doesn't conflict with the other changes.